boot/zephyr/include/flash_map/flash_map.h

```diff
@@ -43,14 +43,64 @@ extern "C" {
  */
 #include <inttypes.h>
 
+/**
+ * @brief Structure describing an area of memory-mapped flash.
+ *
+ * Each flash area has an ID, and contains one or more sectors.
+ *
+ * Multiple flash devices may be available in the system, each of
+ * which may have its own areas. For this reason, flash areas track
+ * which flash device they are part of.
+ */
 struct flash_area {
+    /**
+     * This flash area's ID; unique in the system.
+     */
     uint8_t fa_id;
+
+    /**
+     * ID of the flash device this area is a part of.
+     */
     uint8_t fa_device_id;
+
     uint16_t pad16;
+
+    /**
+     * This area's offset, relative to its flash device's base
+     * address.
+     */
     uint32_t fa_off;
+
+    /**
+     * This area's size, in bytes.
+     */
     uint32_t fa_size;
 };
 
+/**
+ * @brief Structure describing a sector within a flash area.
+ *
+ * Each sector has an offset relative to the start of its flash area
+ * (NOT relative to the start of its flash device), and a size. A
+ * flash area may contain sectors with different sizes.
+ */
+struct flash_sector {
+    /**
+     * Offset of this sector, from the start of its flash area (not device).
+     */
+    uint32_t fs_off;
+
+    /**
+     * Size of this sector, in bytes.
+     */
+    uint32_t fs_size;
+};
+
+/*
+ * Retrieve a flash device's base address.
+ */
+int flash_device_base(uint8_t fd_id, uint32_t *ret);
```

Comment:
This needs to be able to fail, for the non-memory-mapped flash.  At least describe what the return code would be in that case.  Realistically, nobody should be configuring slot 0 as non memory-mapped, but it'd make debugging easier to catch that case.